### PR TITLE
build(actions): revert application version upon e2e failure

### PR DESCRIPTION
**Pull request by stanleynguyen**
## Problem

[Ticket](https://www.notion.so/opengov/Postman-CI-CD-E2E-Rolling-Back-Was-Canary-RFC-ee42f3356a264cdf8832649bb988b569)

## Solution

As stated in the ticket's comment.
To be followed up with another PR for notifying on Slack on build successes & failures

## Deployment Checklist

_Any tasks that need to be done on the environment before releasing of this PR (e.g. Setting environment variables, running migrations, etc). If there's no task to be done, put "N/A"_

- [ ] N/A

### CODE REVIEW

File: .github/workflows/deploy-frontend.yml [hunk 1]
```diff
@@ -16,14 +25,21 @@ jobs:
...
           fi
         id: extract_branch
+      - name: Generate revert command
+        id: generate-revert-command
+        env:
+          BRANCH: ${{ steps.extract_branch.outputs.branch }}
+        run: |
+          aws amplify list-jobs --app-id $APP_ID --branch-name $BRANCH --max-items 1 | jq '.jobSummaries[0]' | echo
+          JOB_ID=$(aws amplify list-jobs --app-id $APP_ID --branch-name $BRANCH --max-items 1 | jq '.jobSummaries[0].jobId')
+          echo "revert_cmd=aws amplify start-job --app-id $APP_ID --branch-name $BRANCH --job-id $JOB_ID --job-type RETRY" >> $GITHUB_OUTPUT
```
Comment (jhibberd-gov) [inline .github/workflows/deploy-frontend.yml:40]: I'm new to GitHub Actions workflows but my understanding is that for each component (backend, frontend, etc.) we're generating a revert command by querying the current code revision deployed for each component. I wonder whether doing this on a component-by-component basis could cause inconsistencies, where we accidentally revert different components to different code revisions. Is there any way we can query for the last revision successfully deployed, and generate the revert commands based on that, to ensure all components will be reverted to the same revision.

NB There's a reasonable chance I misunderstood the code, and this isn't an issue.

File: .github/workflows/deploy-worker.yml [hunk 2]
```diff
@@ -44,11 +52,19 @@ jobs:
...
     runs-on: ubuntu-latest
     steps:
+      - name: Set environment variables
+        run: |
+          if [$CURRENT_ENV == 'production']
+          then
+            echo "SECRET_ID=env/eb/postmangovsg-production-amz2" >> $GITHUB_ENV
+          else
+            echo "SECRET_ID=env/eb/postmangovsg-staging-amz2" >> $GITHUB_ENV
+          fi
```
[the remark this answers is not in the document]
Comment (stanleynguyen) [reply .github/workflows/deploy-worker.yml:67]: Actually for our case yes, but it doesn't represent the correct nature of these IDs imo

File: .github/workflows/deploy-eb.yml [hunk 3]
```diff
@@ -90,35 +98,35 @@ jobs:
...
           sed -i -e "s|@REPO|$ECR_REPOSITORY|g" Dockerrun.aws.json 
           docker build -f Dockerfile -t $ECR_REPOSITORY:$IMAGE_TAG .
-          docker tag $ECR_REPOSITORY:$IMAGE_TAG $ECR_REPOSITORY:$BRANCH
+          docker tag $ECR_REPOSITORY:$IMAGE_TAG $ECR_REPOSITORY:$CURRENT_ENV
           docker push -a $ECR_REPOSITORY
           zip -r "$IMAGE_TAG.zip" .ebextensions .platform Dockerrun.aws.json
 
       - name: Copy to S3
         env:
-          BUCKET_NAME: ${{ secrets.EB_BUCKET_NAME }}
```
[the remark this answers is not in the document]
Comment (stanleynguyen) [reply .github/workflows/deploy-eb.yml:113]: Yup it's actually a refactor that's bundled into this PR as well

File: .github/workflows/deploy-frontend.yml (hunk 1 above)
Comment (stanleynguyen) [reply .github/workflows/deploy-frontend.yml:40]: yup I share this concern too, but right now there's no way to figure out `last successful deployment` that's why I'm resorting to this

Comment (stanleynguyen) [reply .github/workflows/deploy-frontend.yml:40]: But I dont think it's a big concern cos we will be notified of failed deployments and will revert the code before the next deployment attempt

**Comment by jhibberd-gov** [comment]
Thanks for responding to all my comments!

**Review by jhibberd-gov — APPROVED** [review]

**Pull request merged**